src/gst.c

```diff
@@ -0,0 +1,218 @@
+// GStreamer RTP streaming implementation
+
+#include <signal.h>
+#include <pthread.h>
+#include "cloudcam/cloudcam.h"
+#include "cloudcam/log.h"
+#include "cloudcam/gst.h"
+
+// creates a jpeg snapshot of the next available frame
+// buffer returned in *data must be released via free()
+size_t gst_get_jpeg_snapshot(gst_thread_ctx *ctx, void **data)
+{
+  pthread_mutex_lock(&ctx->snapshot_mutex);
+  unsigned long prev_snapshot_frame = ctx->snapshot_frame;
+  // set the update flag and wait until the snapshot is created
+  ctx->snapshot_update_flag = 1;
+  do {
+    pthread_cond_wait(&ctx->snapshot_cond, &ctx->snapshot_mutex);
+  }
+  while (ctx->snapshot_frame == prev_snapshot_frame); // wait until snapshot was actually updated
+  // copy the snapshot and return
+  size_t snapshot_size = ctx->snapshot_size;
+  if (snapshot_size != 0) {
+    *data = malloc(snapshot_size);
+    memcpy(*data, ctx->snapshot_buf, snapshot_size);
+  }
+  else {
+    *data = NULL;
+  }
+  pthread_mutex_unlock(&ctx->snapshot_mutex);
+  return snapshot_size;
+}
+
+// this pad probe passes the video frame to jpeg encoder or drops it depending on if a snapshot was requested
+// also keeps the frame count in ctx->current_frame
+static GstPadProbeReturn jpeg_sink_pad_probe_cb(GstPad          *pad,
+                                                GstPadProbeInfo *info,
+                                                gpointer         user_data)
+{
+  gst_thread_ctx *ctx = (gst_thread_ctx *)user_data;
+  GstPadProbeReturn probe_return = GST_PAD_PROBE_DROP; // by default, drop the buffer
+  pthread_mutex_lock(&ctx->snapshot_mutex);
+  ++ctx->current_frame; // update the total frame count
+  // check if a snapshot was requested and clear the flag if so
+  if (ctx->snapshot_update_flag == 1) {
+    ctx->snapshot_update_flag = 0;
+    probe_return = GST_PAD_PROBE_OK; // pass the buffer to jpeg encoder
+  }
+  pthread_mutex_unlock(&ctx->snapshot_mutex);
+  return probe_return;
+}
+
+// this pad probe receives the jpeg-encoded snapshot (if it was requested and allowed through by jpeg_sink_pad_probe_cb)
+static GstPadProbeReturn jpeg_src_pad_probe_cb(GstPad          *pad,
+                                               GstPadProbeInfo *info,
+                                               gpointer         user_data)
+{
+  gst_thread_ctx *ctx = (gst_thread_ctx *)user_data;
+  GstMapInfo map;
+  GstBuffer *buffer = GST_PAD_PROBE_INFO_BUFFER(info);
+  if (gst_buffer_map(buffer, &map, GST_MAP_READ)) {
```

Comment:
might be a little cleaner to return early instead of put the whole function in a conditional